Add Plugin GroovyDebugX IDE

SunilPharswan opened the pull request:
__GroovyDebuggerX IDE__ plugin, a powerful debugging tool that streamlines Groovy script development in SAP Cloud Platform Integration (CPI) by enabling seamless data transfer between CPI traces and external Groovy IDE.

### Key Features

- __Visual Step Highlighting__: Automatically highlights Groovy script steps that have trace data in the integration flow designer
- __Interactive Debugging__: Click on highlighted Groovy steps to view comprehensive debug information
- __Comprehensive Debug Data__: Multi-tab interface displaying Properties, Headers, Body, Script, Log, and Info data
- __Lazy Loading__: Efficient loading of large trace data only when tabs are accessed
- __External IDE Integration__: One-click "Debug Externally" button to send debug data to Groovy WebIDE (groovyide.com)
- __Cross-Platform Support__: Works with both Neo and Cloud Foundry CPI platforms

### How It Works

1. __Activation__: Click the `{}` button in the message sidebar when viewing trace-mode messages
2. __Visual Feedback__: Groovy script steps with available trace data are highlighted in green
3. __Debug Access__: Click any highlighted step to open a detailed debug popup
4. __Data Transfer__: Use the "Debug Externally" button to send input data, script content, headers, and properties to an external Groovy IDE for testing

### Technical Implementation

- __Plugin Metadata__: Follows CPI-Helper plugin v1.0.0 specification
- __Data Compression__: Uses pako deflateRaw for efficient data transfer to external IDE
- __UI Framework__: Built with Semantic UI for consistent user experience
- __Error Handling__: Comprehensive error handling with user-friendly toast notifications

### Requirements

- __Trace Mode__: Messages must be processed in Trace mode to capture runtime data

### Testing Performed

- ✅ Plugin loads correctly in CPI-Helper extension
- ✅ Button appears in message sidebar for trace-enabled messages
- ✅ Visual highlighting works for Groovy steps with trace data
- ✅ Debug popup displays all required tabs and data
- ✅ Lazy loading functions correctly for large datasets
- ✅ "Debug Externally" button successfully transfers data to Groovy WebIDE
- ✅ Error handling tested with appropriate user feedback

### Files Changed

- `plugins/groovyDebugger.js` - New plugin implementation (700+ lines)
- `manifest.json` - Added plugin to content scripts array
- `lib/pako/pako_deflate.min.js` - For data compression

### Plugin Metadata

"metadataVersion": "1.0.0",
"id": "groovyDebugger", 
"name": "GroovyDebugX IDE",
"version": "1.0.0",
"author": "Sunil Pharswan",
"email": "sunilpharswan4198@gmail.com", 
"website": "https://linkedin.com/in/sunilph",
"description": "GroovyDebugX streamlines Groovy debugging by automating runtime trace extraction. With visual step highlighting and one-click data transfer to Groovy WebIDE, it eliminates manual data entry and accelerates your integration development."

### Checklist

- [x] Plugin follows CPI-Helper plugin specification v1.0.0
- [x] Plugin registered in manifest.json
- [x] Code follows project coding standards
- [x] Comprehensive error handling implemented
- [x] User documentation provided
- [x] Plugin metadata complete and accurate

This plugin addresses the common pain point of debugging Groovy scripts in CPI by providing a seamless bridge between runtime trace data and development environments, significantly improving developer productivity.

SunilPharswan committed:
feat(manifest): add pako library for compression and groovy debugger plugin

- Added "/lib/pako/pako_deflate.min.js" to enable data compression features
- Added "/plugins/groovyDebugger.js" to provide debugging tools for Groovy scripts
- Fixed path for markdown-it library by adding leading slash for consistency

SunilPharswan committed:
style(groovyDebugger): capitalize toast message types for UI consistency

- Updated all showToast calls to use capitalized type strings (e.g., "success" to "Success", "error" to "Error", "warning" to "Warning")
- Ensures consistent capitalization in toast notifications for better user experience and adherence to UI style guidelines.

SunilPharswan committed:
feat(groovyDebugger): highlight only Groovy steps with trace data

Enhanced the Groovy Debugger to fetch trace information and highlight only executed Groovy steps, improving debugging precision by filtering out non-executed elements and updating UI messages for clarity.

SunilPharswan committed:
feat(groovyDebugger): Enhance resetGroovyHighlighting to remove click handlers

- Updated the resetGroovyHighlighting function to also remove click event handlers and cursor styles from all BPMN shape elements, ensuring a complete cleanup after highlighting. This prevents lingering interactions and improves the reset behavior.

SunilPharswan committed:
refactor(groovyDebugger): simplify iFlow URL construction by using direct artifactId API

Remove getIFlowUrl function and replace with direct URL building using pluginHelper.artifactId and tenant. Update script fetching to clean path prefixes for API compatibility, improving reliability and reducing complexity.

SunilPharswan committed:
feat(groovyDebugger): add Log and Info tabs for enhanced debugging

- Add new Log and Info tabs to display run step properties and execution details
- Include helper functions formatLogContent and formatInfoContent for table formatting
- Fetch and integrate runStepData from API for logs and step metadata (start/end times, durations, IDs)
- Swap Properties and Body tab order, making Properties the default active tab
- Improves debugging by providing comprehensive step execution information and results logging

SunilPharswan committed:
feat(groovyDebugger): implement lazy loading for debug content tabs

Refactor createGroovyDebugContent to load Body, Headers, and Script content asynchronously on tab activation via API calls, improving performance by avoiding upfront data fetching. Remove redundant script loading logic from setupGroovyClickHandlers.

SunilPharswan committed:
feat: add direct API call for artifactId retrieval in Groovy debugger

- Introduced getArtifactIdDirectly() function to fetch artifactId via API calls for both Neo and CF platforms, ensuring reliable access when pluginHelper.artifactId is unavailable.
- Updated plugin.onActivated to use the new function, replacing direct usage of pluginHelper.artifactId for better robustness in iFlow data fetching.

SunilPharswan committed:
refactor(groovyDebugger): comment out debug console logs for cleaner code

Commented out unnecessary console.log statements for pluginHelper and runInfo to reduce noise in production logs and improve code cleanliness.

SunilPharswan committed:
refactor(groovyDebugger): remove URL parsing and platform detection logic

Removed redundant code in getArtifactIdDirectly function that extracted integrationFlowId from URL, determined CPI platform, and set urlExtension, as this logic is now handled elsewhere for better modularity.

SunilPharswan committed:
Merge pull request #1 from SunilPharswan/newdev

Newdev

SunilPharswan committed:
feat(groovy-debugger): use formatTrace for script display to enable editor functionality

Refactor script content rendering in createGroovyDebugContent to utilize formatTrace function, providing full editor features like the Body section, instead of plain HTML divs. This improves user experience by allowing interactive editing and better formatting of Groovy scripts in the debugger.

dbeck121 commented:
hey @SunilPharswan 

I will check next week :-) thanks in advance. Looks interesting :-)

SunilPharswan committed:
docs(groovyDebugger): add comprehensive JSDoc documentation to plugin functions

- Added detailed JSDoc comments to global functions, async functions, and helper methods
- Included typedef for plugin configuration object
- Improved code readability and maintainability for the Groovy debugger plugin

dbeck121 commented:
@SunilPharswan 

I checked your plugin and it is very impressive :-) Great Job. Thanks for the contribution!

Only one thing that I need to add it to the general source code:

Could you add a confirm dialog before sending it to groovy ide? I want everyone to be sure that sensible informations might be send to an external service :)

dbeck121 committed:
Merge branch 'dev' into dev

dbeck121 commented:
The codebase was changed so I updated your branch to newest version.

dbeck121 commented:
Hi @SunilPharswan 

maybe one small suggestion:

You are using the groovyide. Maybe could you make the url editable? You can easily add text fields in plugin settings. ( https://github.com/dbeck121/CPI-Helper-Chrome-Extension/blob/dev/docs/readme/PluginREADME.md)

I guess other ides like the https://ide.contiva.com/cpi might also be interested to add the same share functions in the future :-)

SunilPharswan committed:
feat(groovyDebugger): Add confirmation dialog and configurable IDE URL

- Introduce user confirmation prompt before sending debug data to external IDE, enhancing privacy and security by displaying data types and destination domain.
- Add configurable 'externalIdeUrl' setting in plugin settings for custom IDE endpoints, with fallback to default URL.
- Update sendToExternalIDE function to utilize the configurable URL from settings.

SunilPharswan commented:
Hello @dbeck121 ,

Thank you for your feedback.
As suggested a user confirmation dialog box is added before sending data to external ide and configurable 'externalIdeUrl' is added to plugin settings.
Please review and let me you if you have more suggestion.

SunilPharswan committed:
Added : Create custom confirmation popup

SunilPharswan committed:
feat: enhance Groovy Debugger transfer dialog with privacy notices and data labels

Add a privacy notice warning about sensitive business information in transferred data. Reorder dialog elements for better flow and add descriptive labels to checkboxes indicating data types (e.g., sensitive data, source code) to improve user awareness and transparency.

SunilPharswan committed:
fix(groovyDebugger): correct typo in headers label

dbeck121 committed:
Fix bug with external url

dbeck121 commented:
Great! I fixed a small issue but Thanks for your great work. Will be part of next release 3.23
